x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,391 @@
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	testconstants "github.com/zigbee-alliance/distributed-compliance-ledger/integration_tests/constants"
+	testkeeper "github.com/zigbee-alliance/distributed-compliance-ledger/testutil/keeper"
+	sdktestdata "github.com/zigbee-alliance/distributed-compliance-ledger/testutil/testdata"
+	dclauthtypes "github.com/zigbee-alliance/distributed-compliance-ledger/x/dclauth/types"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/keeper"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/types"
+)
+
+type TestSetup struct {
+	T *testing.T
+	// Cdc         *amino.Codec
+	Ctx           sdk.Context
+	Wctx          context.Context
+	Keeper        *keeper.Keeper
+	DclauthKeeper *DclauthKeeperMock
+	UpgradeKeeper *UpgradeKeeperMock
+	Handler       sdk.Handler
+	// Querier     sdk.Querier
+	Plan types.Plan
+}
+
+func (setup *TestSetup) AddAccount(
+	accAddress sdk.AccAddress,
+	roles []dclauthtypes.AccountRole,
+) {
+	for _, role := range roles {
+		setup.DclauthKeeper.On("HasRole", mock.Anything, accAddress, role).Return(true)
+	}
+	setup.DclauthKeeper.On("HasRole", mock.Anything, accAddress, mock.Anything).Return(false)
+}
+
+func (setup *TestSetup) ScheduleUpgradeBehavior(
+	plan types.Plan,
+	err error,
+) {
+	setup.UpgradeKeeper.On("ScheduleUpgrade", mock.Anything, plan).Return(err)
+}
+
+func Setup(t *testing.T) TestSetup {
+	dclauthKeeper := &DclauthKeeperMock{}
+	upgradeKeeper := &UpgradeKeeperMock{}
+	keeper, ctx := testkeeper.DclupgradeKeeper(t, dclauthKeeper, upgradeKeeper)
+
+	trustee := sdktestdata.GenerateAccAddress()
+	plan := testconstants.Plan
+	setup := TestSetup{
+		T:             t,
+		Ctx:           ctx,
+		Wctx:          sdk.WrapSDKContext(ctx),
+		Keeper:        keeper,
+		DclauthKeeper: dclauthKeeper,
+		UpgradeKeeper: upgradeKeeper,
+		Handler:       NewHandler(*keeper),
+		Plan:          plan,
+	}
+
+	setup.AddAccount(trustee, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.ScheduleUpgradeBehavior(plan, nil)
+	return setup
+}
+
+func TestHandler_ProposedUpgradeExists(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress2 := sdktestdata.GenerateAccAddress()
+	trusteeAccAddress3 := sdktestdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// propose new upgrade
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+
+	// propose the same upgrade
+	_, err = setup.Handler(setup.Ctx, msgProposeUpgrade)
```

Comment:
It will be better to verify the repeated proposal from another account.

Reply:
fixed